Group small files into chunks for delta backups [BF-595]

alexole opened the pull request:
<!-- All contributors please complete these sections, including maintainers -->
# About this change - What it does

Adds possibility to chunk small files together into bigger ones when doing delta backup in order to safe costs and improve the speed of backup and recovery.

# Why this way

Usually when doing a backup there are many small files with the size varying between several KBs to several MBs, which sum up to insignificant cumulative size in comparison with the total size of DB.
It does not make much sense in this case to waste resources on calculating hashes of those files and what is worse - spend time on uploading those to the object storage separately (adds on latency for every file) and burn money on every request. Delta functionality is not efficient for those.
Suggestion is to just chunk those files into archives (pretty much the same behaviour as for local-tar) and apply delta logic for files of significant enough size.

Note:
* Some of the type annotations are missing in the old code I moved, as adding those will require some bits of refactoring, which I would like to avoid for now. So this will be done later I hope.

alexole committed:
Move basebackup files into a separate folder

alexole committed:
Bump rohmu version which supports files grouping

alexole commented:
Found some issues when testing, not sure if related to my changes, will move it to draft while investigating. Anyway feel free to review.

alexole committed:
Reduce retry interval right after delta backup failure

kmichel-aiven commented on pghoard/restore.py:
```diff
@@ -82,10 +83,19 @@ def __init__(
     ):
         self.data: bytes = data
-        super().__init__(size=size, new_name=new_name, metadata=metadata, file_type=file_type)
+        self.metadata = metadata
+        super().__init__(size=size, new_name=new_name, file_type=file_type)
+
+    def __eq__(self, other):
+        return isinstance(other, FileDataInfo) and self.data == other.data and self.size == other.size \
+               and self.new_name == other.new_name and self.metadata == other.metadata \
+               and self.file_type == other.file_type
+
+    def __repr__(self):
+        return f"<data> (size={self.size}, file_type={self.file_type}, metadata={self.metadata})"
```
Theses classes with manual `__eq__` and `__repr__` implementations should be dataclasses instead.

alexole replied:
There is still a limitation in dataclasses inheritance, if you have fields with default values in the base class, then the child class should have all fields with default values. Does not fit very nice.

alexole replied:
Let me check if I can workaround this.

alexole replied:
OKish workaround exists from Python 3.10 only (kw_only=True)

kmichel-aiven commented on pghoard/pghoard.py:
```diff
@@ -364,6 +366,16 @@ def _get_delta_basebackup_files(self, site, storage, metadata, basebackup_name_t
             if backup_name != basebackup_name_to_delete:
                 # Keep data file in case if there is still a reference from other backups
                 keep_hexdigests |= backup_hexdigests
+            else:
+                # Add bundles to remove
+                for chunk in meta.get("chunks", []):
+                    basebackup_data_files.append(
+                        os.path.join(
+                            self._get_site_prefix(site),
+                            FileTypePrefixes[FileType.Basebackup_delta_chunk],
+                            chunk["chunk_filename"],
+                        )
+                    )
```
Codecov complains about coverage here, and this looks like something that should particularly be tested.

kmichel-aiven reviewed (dismissed):
Many return types should be dataclass (all the large tuples) to help documentation and avoid swapping-parameters mistakes but I think all of them were pre-existing and just made evident by the extra type hints, this can be improved later.

alexole commented on test/test_restore.py:
```diff
@@ -495,3 +497,82 @@ def check_sha256(cls, fn, expected_sha256):
+                },
+            },
+        }
+        r.storage = Mock()  # pylint: disable=protected-access
```
Leftover

alexole commented on test/test_restore.py:
```diff
@@ -495,3 +497,82 @@ def check_sha256(cls, fn, expected_sha256):
             assert actual_sha256.hexdigest() == expected_sha256
+
+
+class TestRestore:
```
It is unnecessary, I just wanted to make it visible, that this class is testing methods of Restore class, i.e. `get_delta_basebackup_data`

alexole committed:
Small changes addressing review comments, mainly type annotations

rikonen reviewed (dismissed):
Didn't spot any functional issues except for likely confusing metrics that could be improved.

I'm not a big fan of `__init__.py` files with big implementation and would consider moving the basebackup stuff from that to a different file.

It would be good to ensure the latest main version without these changes works normally as part of the bigger solution before merging these to avoid having huge diff when switching to main version.

alexole committed:
Delta backups: chunk small files

Usually when doing a backup there are many small files with the size
varying between several KBs to several MBs, which sum up to insignificant
cumulative size in comparison with the total size of DB.
It does not make much sense in this case to waste resources on calculating
hashes of those files and what is worse - spend time on uploading those to
the object storage separately (adds on latency for every file) and burn
money on every request. Delta functionality is not efficient for those.
Suggestion is to just chunk those files into archives (pretty much the same
behaviour as for local-tar) and apply delta logic for files of significant
enough size.

alexole committed:
Reduce size of delta chunk to 500 MB

alexole committed:
Adjust progress metric, so it's correct for delta backups too

alexole committed:
Move basebackup code out of init file

alexole committed:
Test delta backups mode with different parameters

rikonen commented on pghoard/basebackup/chunks.py:
```diff
@@ -0,0 +1,327 @@
+            for task in pending_compress_and_encrypt_tasks:
+                chunk_files.append(task.result())
+                self.metrics.gauge(
+                    "pghoard.basebackup_estimated_progress",
```
This comment was resolved but it is still sending `pghoard.basebackup_estimated_progress` metric data point for the delta chunk uploads too, which seems confusing. For that context the metric name should perhaps be `pghoard.delta_chunks_estimated_progress` or something like that to clarify what the context is.

alexole replied:
I wanted a unified metric, so no need to do any changes for the dashboard.
Also `pghoard.basebackup_estimated_progress` even for delta will reflect the overall progress, no only chunks. So if we have 2 chunks an d 2 delta files, the `pghoard.basebackup_estimated_progress` will finish at 50% when uploading chunks and the remaining 50% will be left for delta files.

rikonen replied:
Ah, I see the target size is given so that this should actually correctly reflect the total, nevermind

rikonen commented on pghoard/basebackup/delta.py:
```diff
@@ -0,0 +1,479 @@
+        skip_upload = False
+        start_time = time.monotonic()
+
+        input_size = file_obj.seek(0, os.SEEK_END)
```
Existing issue but I guess file size can change while it's being uploaded and instead of reading the size here we should check how many bytes we actually ended up uploading (and if hash matches then checking what's the actual size of the file that matched)?

rikonen reviewed (dismissed):
Changes seem fine to me. The comment about file size being potentially wrong is probably worth checking but since this doesn't change the logic shouldn't be a blocker for merging this.

rdunklau commented on pghoard/basebackup/chunks.py:
```diff
@@ -0,0 +1,328 @@
+    @staticmethod
+    def chunk_path_to_middle_path_name(chunk_path: Path, file_type: FileType) -> Tuple[Path, str]:
+        chunk_rel_path = chunk_path.relative_to(chunk_path.parent.parent)
+        if file_type == FileType.Basebackup_chunk:
```
You should probably use pghoard.common.FileTypePrefixes here to avoid repeating those constants.

rdunklau commented on pghoard/basebackup/delta.py:
```diff
@@ -0,0 +1,479 @@
+                    self.log.info("Cleaning up already uploaded new backup files")
+
+                    for new_hash in new_hashes:
+                        key = os.path.join(self.site_config["prefix"], "basebackup_delta", new_hash)
```
Same here, we should avoid those "magic strings" and use the FileTypePrefixes dict.

alexole committed:
Remove code duplication

alexole commented:
Addressed small change request @rdunklau

alanfranz-at-work reviewed (approved):
This already got a good amount of feedback; it looks good to me and it seems that comments were addressed, tests are passing and coverage is going up, so I'm all-in for merging.
